### PR TITLE
Working to support Gaia Rho

### PR DESCRIPTION
**Pull request by soareschen**
## Description

This is a work in progress to track support for [gaia/rho-ibc](https://github.com/cosmos/gaia/tree/rho-ibc), with the `gaiad` binary provided from https://github.com/informalsystems/cosmos.nix/pull/86.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: relayer/src/chain/cosmos/wait.rs [hunk 1]
```diff
@@ -1,114 +1,117 @@
...
+        });
+        join_handles.push(handle);
+    }
+
+    let mut responses = Vec::new();
+
+    for handle in join_handles.into_iter() {
+        let response = handle.await.map_err(Error::join)??;
+        responses.push(response);
+    }
```
Comment (mzabaluev) [inline relayer/src/chain/cosmos/wait.rs:72]: It's more efficient to use `FuturesOrdered`.

Comment (soareschen) [reply relayer/src/chain/cosmos/wait.rs]: Thanks for the suggestions! Loops are easier to write during quick hacks, though I have found the time to refactor them to use iterators and streams.

File: relayer/src/chain/cosmos/wait.rs [hunk 2]
```diff
@@ -50,26 +50,16 @@ pub async fn wait_tx_hashes(
...
+    let tasks = tx_hashes
+        .iter()
+        .map(|tx_hash| async { wait_tx_hash(rpc_client, rpc_address, timeout, tx_hash).await })
+        .collect::<FuturesOrdered<_>>();
+
+    let responses = tasks
+        .collect::<Vec<_>>()
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()?;
```
Comment (mzabaluev) [inline relayer/src/chain/cosmos/wait.rs:62]: A bit hard to follow the logic here. The first `.collect` makes sure all futures are executed, whereas `.try_collect` would bail out after one of them resolved an error. Then, the resolved vector of `Result` from all tasks is transformed to determine if the overall result is a success, anyway, and propagate the first error found. Would it be simpler to just use `.try_collect` here?

**Review by mzabaluev — COMMENTED** [review]
Good use of future combinators!

File: relayer/src/chain/cosmos/wait.rs (hunk 2 above)
Comment (soareschen) [reply relayer/src/chain/cosmos/wait.rs]: You are right, I managed to simplify it further!
